### PR TITLE
[ODS-6510] Apply DRY principle to security/authorization logic (through consolidation of common code)

### PR DESCRIPTION
**Pull request by gmcelhanon**
⚠️**NOTE TO REVIEWER:** See comment on https://github.com/Ed-Fi-Alliance-OSS/Ed-Fi-ODS-Implementation/pull/1147 for a high-level overview of the changes made in this branch.

### CODE REVIEW

File: Application/EdFi.Ods.Api/Security/Authorization/AuthorizationBasis/AuthorizationStrategyResolver.cs [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+                    {
+                        var authorizationStrategy = authorizationStrategyProvider.GetByName(strategyName);
+
+                        if (authorizationStrategy != null)
+                        {
+                            return authorizationStrategy;
+                        }
+                    }
+
+                    throw new SecurityConfigurationException(
```
Comment (mjaksn) [inline Application/EdFi.Ods.Api/Security/Authorization/AuthorizationBasis/AuthorizationStrategyResolver.cs:46]: Would it increase the helpfulness of the exception message here to collect the names of any authorizationStrategies that can't be found and throwing an exception which includes them all, rather than stopping evaluation when on the first on that can't be found?

File: Application/EdFi.Ods.Api/Security/Authorization/EntityAuthorization/EntityAuthorizationSqlBuilder.cs [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+namespace EdFi.Ods.Api.Security.Authorization.EntityAuthorization;
+
+public interface IEntityAuthorizationSqlBuilder
+{
+    string BuildExistenceCheckSql(AuthorizationStrategyFilterResults[] resultsWithPendingExistenceChecks);
+}
+
+public class EntityAuthorizationSqlBuilder : IEntityAuthorizationSqlBuilder
+{
+    public string BuildExistenceCheckSql(AuthorizationStrategyFilterResults[] resultsWithPendingExistenceChecks)
```
Comment (mjaksn) [inline Application/EdFi.Ods.Api/Security/Authorization/EntityAuthorization/EntityAuthorizationSqlBuilder.cs:23]: Should unit tests be added to explicitly cover this implementation, or is its functionality sufficiently covered by other unit/integration tests?

File: Application/EdFi.Ods.Api/Security/Authorization/EntityAuthorization/EntityAuthorizer.cs [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+namespace EdFi.Ods.Api.Security.Authorization.EntityAuthorization;
+
+public class EntityAuthorizer : IEntityAuthorizer
+{
+    private readonly IDataManagementAuthorizationPlanFactory _dataManagementAuthorizationPlanFactory;
+    private readonly IEntityInstanceDelegateFilterAuthorizer _entityInstanceDelegateFilterAuthorizer;
+    private readonly IEntityInstanceAuthorizationValidator _entityInstanceAuthorizationValidator;
+    private readonly IEntityInstanceViewBasedFilterAuthorizer _entityInstanceViewBasedFilterAuthorizer;
+
+    private readonly ILog _logger = LogManager.GetLogger(typeof(EntityAuthorizer));
```
Comment (mjaksn) [inline Application/EdFi.Ods.Api/Security/Authorization/EntityAuthorization/EntityAuthorizer.cs:26]: This logger is unused and could be removed.

File: Application/EdFi.Ods.Tests/EdFi.Ods.Security/Authorization/EntityAuthorization/EntityInstanceViewBasedFilterAuthorizerTests.cs [hunk 4]
```diff
@@ -0,0 +1,571 @@
...
+                {
+                    new FilterAuthorizationResult
+                    {
+                        FilterContext = new AuthorizationFilterContext
+                        {
+                            SubjectEndpointNames = new[] { "TestParam" },
+                            SubjectEndpointValues = new object[] { 456 },
+                            // Custom view-based authorization will have no claim values
+                            // ClaimParameterName = "ClaimParam",
+                            // ClaimEndpointValues = [123]
```
Comment (mjaksn) [inline Application/EdFi.Ods.Tests/EdFi.Ods.Security/Authorization/EntityAuthorization/EntityInstanceViewBasedFilterAuthorizerTests.cs:391]: Can this commented out code be removed?

File: Application/EdFi.Ods.Tests/EdFi.Ods.Security/Authorization/EntityAuthorization/EntityInstanceViewBasedFilterAuthorizerTests.cs [hunk 5]
```diff
@@ -0,0 +1,571 @@
...
```
Comment (mjaksn) [inline Application/EdFi.Ods.Tests/EdFi.Ods.Security/Authorization/EntityAuthorization/EntityInstanceViewBasedFilterAuthorizerTests.cs:449]: Can this commented out code be removed?

File: Application/EdFi.Ods.Tests/EdFi.Ods.Security/Authorization/EntityAuthorization/EntityInstanceViewBasedFilterAuthorizerTests.cs [hunk 6]
```diff
@@ -0,0 +1,571 @@
...
```
Comment (mjaksn) [inline Application/EdFi.Ods.Tests/EdFi.Ods.Security/Authorization/EntityAuthorization/EntityInstanceViewBasedFilterAuthorizerTests.cs:508]: Can this commented out code be removed?

File: Application/EdFi.Ods.Tests/EdFi.Ods.Security/Authorization/EdFiAuthorizationProviderTests.cs [hunk 7]
```diff
@@ -36,8 +36,9 @@
 using Shouldly;
 using Test.Common;
 using Action = EdFi.Security.DataAccess.Models.Action;
-using Helper = EdFi.Ods.Tests.EdFi.Ods.Api.Security.Authorization.AuthorizationTestsHelper;
-
+// using Helper = EdFi.Ods.Tests.EdFi.Ods.Api.Security.Authorization.AuthorizationTestsHelper;
+using ValidationResult = System.ComponentModel.DataAnnotations.ValidationResult;
+/*
```
Comment (mjaksn) [inline Application/EdFi.Ods.Tests/EdFi.Ods.Security/Authorization/EdFiAuthorizationProviderTests.cs:41]: Just want to double check, was the intent to comment out this entire file? If so, can a TODO be added to describe the anticipated next steps regarding its content?

**Review by mjaksn — CHANGES_REQUESTED** [review]

File: Application/EdFi.Ods.Api/Security/Extensions/EdFiAuthorizationContextExtensions.cs [hunk 8]
```diff
@@ -12,16 +12,33 @@ public static class EdFiAuthorizationContextExtensions
...
     /// Gets one of two possible strings supplied corresponding to the authorization phase of the authorization context.
     /// </summary>
-    /// <param name="authorizationContext">The authorization context which identifies the current authorization phase.</param>
+    /// <param name="dataManagementRequestContext">The authorization context which identifies the current authorization phase.</param>
     /// <param name="existingDataText">The text to return if the phase is <see cref="AuthorizationPhase.ExistingData"/></param>
     /// <param name="proposedDataText">The text to return if the phase is <see cref="AuthorizationPhase.ProposedData"/></param>
     /// <returns>The appropriate supplied string.</returns>
     public static string GetPhaseText(
-        this EdFiAuthorizationContext authorizationContext,
+        this DataManagementRequestContext dataManagementRequestContext,
```
Comment (mjaksn) [inline Application/EdFi.Ods.Api/Security/Extensions/EdFiAuthorizationContextExtensions.cs:20]: This class can be renamed EdFiDataManagementRequestContextExtensions.

File: Application/EdFi.Ods.Tests/EdFi.Ods.Security/Authorization/AuthorizationBasis/RequestEvaluationStrategiesSelectorTests.cs [hunk 9]
```diff
@@ -0,0 +1,217 @@
...
+        // Act
+        var result = _selector.TryGetAuthorizationStrategyNames(requestEvaluation, out var strategyNames);
+
+        // Assert
+        result.ShouldBeTrue();
+        strategyNames.ShouldBe(defaultStrategies1);
+    }
+
+    [Test]
+    public void TryGetAuthorizationStrategyNames_ShouldReturnEmpty_WhenAllStrategiesAreEmpty()
```
[the remark this answers is not in the document]
Comment (gmcelhanon) [reply Application/EdFi.Ods.Tests/EdFi.Ods.Security/Authorization/AuthorizationBasis/RequestEvaluationStrategiesSelectorTests.cs:188]: Renamed.

File: Application/EdFi.Ods.Api/Security/Authorization/AuthorizationBasis/AuthorizationStrategyResolver.cs (hunk 1 above)
Comment (gmcelhanon) [reply Application/EdFi.Ods.Api/Security/Authorization/AuthorizationBasis/AuthorizationStrategyResolver.cs:46]: It's quite an edge case, but it's also low hanging fruit. I've added the behavior and new test coverage.

File: Application/EdFi.Ods.Api/Security/Authorization/EntityAuthorization/EntityAuthorizer.cs (hunk 3 above)
Comment (gmcelhanon) [reply Application/EdFi.Ods.Api/Security/Authorization/EntityAuthorization/EntityAuthorizer.cs:26]: Removed.

File: Application/EdFi.Ods.Tests/EdFi.Ods.Security/Authorization/EntityAuthorization/EntityInstanceViewBasedFilterAuthorizerTests.cs (hunk 5 above)
Comment (gmcelhanon) [reply Application/EdFi.Ods.Tests/EdFi.Ods.Security/Authorization/EntityAuthorization/EntityInstanceViewBasedFilterAuthorizerTests.cs:449]: I've removed the two lines of "code" but kept the comment as it is pertinent to Custom view-based authorization testing for someone looking at the tests.

File: Application/EdFi.Ods.Tests/EdFi.Ods.Security/Authorization/EdFiAuthorizationProviderTests.cs (hunk 7 above)
Comment (gmcelhanon) [reply Application/EdFi.Ods.Tests/EdFi.Ods.Security/Authorization/EdFiAuthorizationProviderTests.cs:41]: I meant to remove this after the decomposition of its behavior into smaller units. Removing now.

File: Application/EdFi.Ods.Api/Security/Extensions/EdFiAuthorizationContextExtensions.cs (hunk 8 above)
Comment (gmcelhanon) [reply Application/EdFi.Ods.Api/Security/Extensions/EdFiAuthorizationContextExtensions.cs:20]: I consolidated the usages to eliminate one of the methods and have renamed this class and file to `AuthorizationPhaseExtensions`.
